File: scafi3-integration/src/test/resources/js/protobuf-exchange/program.mjs
import proto from "./messages.js";

function aggregateProgram(lang) {
    const temperature = senseTemperature();
    return lang.exchange(lang.Field.of(temperature), n =>
        returning(n).send(lang.Field.of(temperature))
    );
}

function senseTemperature() {
    return proto.TemperatureSensor.create({ id: `#${deviceId}`, temperature: deviceId * 10 });
}

proto.TemperatureSensor.prototype.toString = function () {
    return `Sensor{id=${this.id}, temp=${this.temperature.toFixed(2)}}`;
};

File: scafi3-integration/src/test/resources/c/protobuf-exchange/program.c
#include "messages.pb-c.h"
#include "protobuf_codable.c"

// Custom toString function for TemperatureSensor
const signed char* temperature_sensor_to_str(const void* data);

DEFINE_PROTOBUF_MESSAGE(temperature_sensor, temperature_sensor_to_str)

const ProtobufValue* sense_temperature() {
    TemperatureSensor* sensor = malloc(sizeof(TemperatureSensor));
    if (!sensor) return NULL;
    temperature_sensor__init(sensor);
    sensor->id = strdup("#{{ deviceId }}");
    sensor->temperature = {{ deviceId }} * 10.0f;
    return proto_of(sensor, temperature_sensor);
}

const void* aggregate_program(const AggregateLibrary* lang) {
    const Field* temp = lang->Field.of((const BinaryCodable*) sense_temperature());
    return lang->exchange(temp, fn(ReturnSending*, (const Field* f), {
        return return_sending(f, temp);
    }));
}

const signed char* temperature_sensor_to_str(const void* data) {
    if (!data) return NULL;
    const ProtobufValue* pv = data;
    const TemperatureSensor* sensor = (const TemperatureSensor*)pv->message;
    const signed char* str = malloc(128);
    if (str) {
        snprintf((char*)str, 128, "Sensor{id=%s, temp=%.2f}", sensor->id, sensor->temperature);
    }
    return str;
}
